telemetry/tracks/class-tracks-event-queue.php

```diff
@@ -0,0 +1,72 @@
+<?php
+/**
+ * Telemetry: Tracks Pixel class
+ *
+ * @package Automattic\VIP\Telemetry
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\VIP\Telemetry\Tracks;
+
+use WP_Error;
+use function Automattic\VIP\Logstash\log2logstash;
+
+/**
+ * Handles queued events to be sent to the Tracks service.
+ */
+class Tracks_Event_Queue {
+	/**
+	 * @var Tracks_Client
+	 */
+	private $client;
+
+	/**
+	 * Events queued to be sent to the Tracks API.
+	 *
+	 * @var array<Tracks_Event>
+	 */
+	protected $events = array();
+
+	/**
+	 * Constructor. Registers the shutdown hook to record any and all events.
+	 */
+	public function __construct( Tracks_Client $client ) {
+		$this->client = $client;
+
+		// Register the shutdown hook to record any and all events
+		add_action( 'shutdown', array( $this, 'record_events' ) );
+	}
+
+	/**
+	 * Enqueues an event to be recorded asynchronously.
+	 *
+	 * @param Tracks_Event $event The event to record.
+	 * @return bool|WP_Error True if the event was enqueued for recording.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function record_event_asynchronously( Tracks_Event $event ) {
+		$is_event_recordable = $event->is_recordable();
+
+		if ( true !== $is_event_recordable ) {
+			return $is_event_recordable;
+		}
+
+		$this->events[] = $event;
+
+		return true;
+	}
+
+	/**
+	 * Records all queued events synchronously.
+	 */
+	public function record_events(): void {
+		if ( [] === $this->events ) {
+			return;
+		}
+
+		$this->client->batch_record_events( $this->events );
```

Comment:
If there's an error during the recording operation, all the events would be dropped. Wondering if it's worth re-trying based on the status code? If it's a 4xx then no, but if it's a 5xx it might be?

Reply:
What is the cost of a missed event? We need to discuss this because good retries require a back-off strategy, which complicates many things. Even with back-off, the Tracks system might not appreciate retries because they could load the system more when it is already struggling. So, if we want to complicate the system with retries, we have to have a very good reason to do that.

I'm looking at VIP CLI implementation, and there are no retries:

https://github.com/Automattic/vip-cli/blob/2b4f7da6d6e80ecb98a99e7e643ba91d38f8a307/src/lib/analytics/clients/tracks.ts#L119-L136

Reply:
I should have added this in my comment, but I'm 50/50 on it for this reason. Considering it's telemetry, I'd argue that the cost of a missed event is not really worth it. As is we are doing this on shutdown to ensure that there's minimal disruption to a site's operations.

Perhaps it's worth adding this as a comment so the decision behind it is known.

Reply:
Did that 👍